File: pegpy/nezcc/parser.ts
export class ParseTree {
  public tag: string;
  public urn: any;
  public inputs: string;
  public spos: number;
  public epos: number;
  public nodes: [string, ParseTree][];

  public constructor(tag: string, spos: number, epos: number, child: any) {
    this.tag = tag;
    this.urn = child;
    this.inputs = '';
    this.spos = spos;
    this.epos = epos;
    this.nodes = ParseTree.empties;
  }

  static empties: [string, ParseTree][] = []

  protected setup(urn: string, inputs: string) {
    if (this.urn !== null) {
      const nodes: [string, ParseTree][] = [];
      var entry: Merge = this.urn;
      while (entry !== null) {
        nodes.push([entry.edge, entry.child.setup(urn, inputs)])
        entry = entry.prev;
      }
      this.nodes = nodes.reverse();
    }
    this.urn = urn;
    this.inputs = inputs;
    var t: any = this;
    for (var i = 0; i < this.nodes.length; i += 1) {
      t[i] = this.nodes[i][1];
      if (this.nodes[i][0] !== '') {
        t[this.nodes[i][0]] = this.nodes[i][1];
      }
    }
    return this;
  }

  public is(tag: string) {
    return this.tag === tag;
  }

  public isError() {
    return this.tag === 'err';
  }

  public subs() {
    const subs: ParseTree[] = [];
    for (var i = 0; i < this.nodes.length; i += 1) {
      subs.push(this.nodes[i][1]);
    }
    return subs;
  }

  public size() {
    return this.nodes.length;
  }

  public contains(edge: string) {
    for (var i = 0; i < this.nodes.length; i += 1) {
      if (this.nodes[i][0] === edge) return true;
    }
    return false;
  }

  public get(index: any) {
    return (this as any)[index];
  }

  public tokenize(index?: any, defstr?: string) {
    if (index === undefined) {
      return this.inputs.substring(this.spos, this.epos);
    }
    const child = (this as any)[index];
    if (child === undefined) {
      return (defstr || '');
    }
    return child.tokenize();
  }

  private pos(pos: number) {
    const s = this.inputs;
    pos = Math.min(pos, s.length);
    var row = 0;
    var col = 0;
    for (var i = 0; i <= pos; i += 1) {
      if (s.charCodeAt(i) == 10) {
        row += 1;
        col = 0;
      }
      else {
        col += 1;
      }
    }
    return [pos, row, col]
  }

  public begin() {
    return this.pos(this.spos);
  }

  public end() {
    return this.pos(this.spos);
  }

  public length() {
    return this.epos - this.spos;
  }

  public toString() {
    const sb: string[] = [];
    this.strOut(sb);
    return sb.join('');
  }

  protected strOut(sb: string[]) {
    sb.push("[#")
    sb.push(this.tag)
    for (const node of this.nodes) {
      sb.push(node[0] === '' ? ' ' : ` ${node[0]}=`)
      node[1].strOut(sb);
    }
    if (this.nodes.length == 0) {
      const s = this.inputs.substring(this.spos, this.epos);
      sb.push(" '");
      sb.push(s);
      sb.push("'");
    }
    sb.push("]")
  }

}

class Merge {
  public prev: any;
  public edge: string;
  public child: ParseTree;

  public constructor(prev: any, edge: string, child: ParseTree) {
    this.prev = prev;
    this.edge = edge;
    this.child = child;
  }
}

class ParserContext {
  public urn: string;
  public inputs: String;
  public pos: number;
  public epos: number;
  public head_pos: number;
  public ast: any;
  public state: State | null;
  public memos: Memo[];
  public constructor(urn: string, inputs: string, pos: number, epos: number) {
    this.urn = urn;
    this.inputs = inputs;
    this.pos = pos;
    this.epos = epos;
    this.head_pos = pos;
    this.ast = null;
    this.state = null;
    this.memos = []
    for (var i = 0; i < 1789; i++) {
      this.memos.push(new Memo());
    }
  }
}

const EMPTY = (px: ParserContext) => {
  return true;
}

const pEmpty = () => {
  return EMPTY;
}

const ANY = (px: ParserContext) => {
  if (px.pos < px.epos) {
    px.pos += 1;
    return true;
  }
  return false;
}

const pAny = () => {
  return ANY;
}

const pChar = (text: string) => {
  const text_length = text.length;

  return (px: ParserContext) => {
    if (px.inputs.startsWith(text, px.pos)) {
      px.pos += text_length;
      return true;
    }
    return false;
  }
}

const find_codemax = (chars: string, ranges: string[]) => {
  var code = 0;
  for (var i = 0; i < chars.length; i += 1) {
    code = Math.max(chars.charCodeAt(i), code);
  }
  for (const range of ranges) {
    code = Math.max(range.charCodeAt(0), code);
    code = Math.max(range.charCodeAt(1), code);
  }
  return code;
}

const set_bitmap = (bitmap: Uint8Array, c: number) => {
  const n = (c / 8) | 0;
  const mask = 1 << ((c % 8) | 0);
  //console.log(n);
  //console.log(bitmap[n])
  bitmap[n] |= mask;
  //console.log(bitmap);
}

const pRange = (chars: string, ranges: string[]) => {
  const codemax = find_codemax(chars, ranges) + 1;
  const bitmap = new Uint8Array(((codemax / 8) | 0) + 1);
  bitmap[0] = 2;
  for (var i = 0; i < chars.length; i += 1) {
    set_bitmap(bitmap, chars.charCodeAt(i));
  }
  for (const range of ranges) {
    for (var c = range.charCodeAt(0); c <= range.charCodeAt(1); c += 1) {
      set_bitmap(bitmap, c);
    }
  }
  //console.log(`bitmap ${ bitmap.length } `)
  return (px: ParserContext) => {
    if (px.pos < px.epos) {
      const c = px.inputs.charCodeAt(px.pos);
      const n = (c / 8) | 0;
      const mask = 1 << ((c % 8) | 0);
      if (n < bitmap.length && (bitmap[n] & mask) === mask) {
        px.pos += 1;
        return true;
      }
    }
    return false;
  }
}

const pMany = (match: (px: ParserContext) => boolean) => {
  return (px: ParserContext) => {
    var pos = px.pos;
    var ast = px.ast;
    while (match(px) && px.pos > pos) {
      pos = px.pos;
      ast = px.ast;
    }
    px.head_pos = Math.max(px.pos, px.head_pos);
    px.pos = pos;
    px.ast = ast;
    return true;
  }
}

const pMany1 = (match: (px: ParserContext) => boolean) => {
  return (px: ParserContext) => {
    if (match(px)) {
      var pos = px.pos;
      var ast = px.ast;
      while (match(px) && px.pos > pos) {
        pos = px.pos;
        ast = px.ast;
      }
      px.head_pos = Math.max(px.pos, px.head_pos);
      px.pos = pos;
      px.ast = ast;
      return true;
    }
    return false;
  }
}

const pAnd = (match: (px: ParserContext) => boolean) => {
  return (px: ParserContext) => {
    const pos = px.pos;
    if (match(px)) {
      px.head_pos = Math.max(px.pos, px.head_pos);
      px.pos = pos;
      return true;
    }
    return false;
  }
}

const pNot = (match: (px: ParserContext) => boolean) => {
  return (px: ParserContext) => {
    const pos = px.pos;
    const ast = px.ast;
    if (match(px)) {
      px.head_pos = Math.max(px.pos, px.head_pos);
      px.pos = pos;
      px.ast = ast;
      return false;
    }
    return true;
  }
}

const pOption = (match: (px: ParserContext) => boolean) => {
  return (px: ParserContext) => {
    const pos = px.pos;
    const ast = px.ast;
    if (!match(px)) {
      px.head_pos = Math.max(px.pos, px.head_pos);
      px.pos = pos;
      px.ast = ast;
    }
    return true;
  }
}

const pSeq = (...matches: ((px: ParserContext) => boolean)[]) => {
  return (px: ParserContext) => {
    for (const match of matches) {
      if (!match(px)) {
        return false;
      }
    }
    return true;
  }
}

const pSeq2 = (match: (px: ParserContext) => boolean, match2: (px: ParserContext) => boolean) => {
  return (px: ParserContext) => {
    return match(px) && match2(px);
  }
}

const pSeq3 = (match: (px: ParserContext) => boolean, match2: (px: ParserContext) => boolean, match3: (px: ParserContext) => boolean) => {
  return (px: ParserContext) => {
    return match(px) && match2(px) && match3(px);
  }
}

const pOre = (...matches: ((px: ParserContext) => boolean)[]) => {
  return (px: ParserContext) => {
    const pos = px.pos;
    const ast = px.ast;
    for (const match of matches) {
      if (match(px)) {
        return true;
      }
      px.head_pos = Math.max(px.pos, px.head_pos);
      px.pos = pos;
      px.ast = ast;
    }
    return false;
  }
}

const pOre2 = (match: (px: ParserContext) => boolean, match2: (px: ParserContext) => boolean) => {
  return (px: ParserContext) => {
    const pos = px.pos;
    const ast = px.ast;
    if (match(px)) {
      return true;
    }
    px.head_pos = Math.max(px.pos, px.head_pos);
    px.pos = pos;
    px.ast = ast;
    return match2(px);
  }
}


const pRef = (peg: any, name: string) => {
  if (peg[name]) {
    return peg[name];
  }
  return (px: ParserContext) => {
    return peg[name](px)
  }
}

class Memo {
  public key: number;
  public constructor() {
    this.key = -1;
  }
}

const pNode = (match: (px: ParserContext) => boolean, tag: string, shift: number) => {
  return (px: ParserContext) => {
    const pos = px.pos
    px.ast = null;
    if (match(px)) {
      px.ast = new ParseTree(tag, pos + shift, px.pos, px.ast);
      return true;
    }
    return false;
  }
}

//def Merge(prev, edge, child):
//return (prev, edge, child)

const pEdge = (edge: string, match: (px: ParserContext) => boolean) => {
  return (px: ParserContext) => {
    const ast = px.ast;
    if (match(px)) {
      px.ast = new Merge(ast, edge, px.ast);
      return true;
    }
    return false;
  }
}

const pFold = (edge: string, match: (px: ParserContext) => boolean, tag: string, shift: number) => {
  return (px: ParserContext) => {
    const pos = px.pos;
    px.ast = new Merge(null, edge, px.ast);
    if (match(px)) {
      px.ast = new ParseTree(tag, pos + shift, px.pos, px.ast);
      return true;
    }
    return false;
  }
}

const pAbs = (match: (px: ParserContext) => boolean) => {
  return (px: ParserContext) => {
    const ast = px.ast;
    if (match(px)) {
      px.ast = ast;
      return true;
    }
    return false;
  }
}

const pSkipErr = () => {
  return (px: ParserContext) => {
    px.pos = Math.min(px.head_pos, px.epos);
    return true;
  }
}

class State {
  public sid: number;
  public value: any;
  public prev: State | null;
  public constructor(sid: number, value: any, prev: State | null) {
    this.sid = sid;
    this.value = value;
    this.prev = prev;
  }
}

const getstate = (state: State | null, sid: number) => {
  while (state !== null) {
    if (state.sid === sid) {
      return state;
    }
    state = state.prev;
  }
  return null;
}

const pSymbol = (sid: number, match: (px: ParserContext) => boolean) => {
  return (px: ParserContext) => {
    const pos = px.pos
    if (match(px)) {
      px.state = new State(sid, px.inputs.substring(pos, px.pos), px.state)
      return true;
    }
    return false;
  }
}

const pExists = (sid: number) => {
  return (px: ParserContext) => {
    return getstate(px.state, sid) !== null;
  }
}

const pMatch = (sid: number) => {
  return (px: ParserContext) => {
    const state = getstate(px.state, sid);
    if (state !== null) {
      if (px.inputs.startsWith(state.value, px.pos)) {
        px.pos += (state.value as string).length;
        return true;
      }
    }
    return false;
  }
}

const pScope = (match: (px: ParserContext) => boolean) => {
  return (px: ParserContext) => {
    const state = px.state;
    const res = match(px);
    px.state = state;
    return res;
  }
}

export const generate = (start: string) => {
  const match = grammar(start);
  if (match === undefined) {
    console.log(`undefined ${start}`)
    console.log(peg)
  }
  return (inputs: string, options?: any) => {
    const op = (options === undefined) ? {} : options;
    const pos = 0;
    const px = new ParserContext(op['urn'] || '(unknown source)', inputs, 0, inputs.length);
    if (match(px)) {
      if (px.ast === null) {
        px.ast = new ParseTree('', pos, px.pos, null);
      }
    }
    else {
      px.ast = new ParseTree('err', px.head_pos, px.head_pos + 1, null);
    }
    return px.ast.setup(px.urn, inputs);
  }
}

let peg: any = null;

const grammar = (start: string) => {
  if (peg === null) {
    peg = {};
    //TPEG
  }
  return peg[start];
}

const example = (start: string, sample?: string) => {
  const parser = generate(start);
  const t = parser(sample || 'abc');
  console.log(`${start} ${sample}`)
  console.log(t.toString());
}

//EXAMPLE

// pegpy nezcc -g math.tpeg parser.ts > math.ts
// npx ts-node math.ts 

File: pegpy/nezcc/parser.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.function.Function;

class ParTree {
  public ParTree prev;
  public String label;
  public int spos;
  public int epos;
  public ParTree child;

  public ParTree(ParTree prev, String tag, int spos, int epos, ParTree child) {
    this.prev = prev;
    this.label = tag;
    this.spos = spos;
    this.epos = epos;
    this.child = child;
  }
}

class Memo {
  public long key;

  public Memo() {
    this.key = -1;
  }
}

class State {
  public int sid;
  public String value;
  public State prev;

  public State(int sid, String value, State prev) {
    this.sid = sid;
    this.value = value;
    this.prev = prev;
  }

  public static State get(State state, int sid) {
    while (state != null) {
      if (state.sid == sid) {
        return state;
      }
      state = state.prev;
    }
    return null;
  }
}

class ParserContext {
  public String urn;
  public String inputs;
  public int pos;
  public int epos;
  public int head_pos;
  public ParTree ast;
  public State state;
  public Memo[] memos;

  public ParserContext(String urn, String inputs, int pos, int epos) {
    this.urn = urn;
    this.inputs = inputs;
    this.pos = pos;
    this.epos = epos;
    this.head_pos = pos;
    this.ast = null;
    this.state = null;
    this.memos = new Memo[1789];
    for (int i = 0; i < 1789; i++) {
      this.memos[i] = new Memo();
    }
  }
}

interface ParseFunc {
  public boolean apply(ParserContext px);
}

class Parser {
  public static ParseFunc EMPTY = (ParserContext px) -> {
    return true;
  };

  public static ParseFunc pEmpty() {
    return EMPTY;
  }

  public static ParseFunc ANY = (ParserContext px) -> {
    if (px.pos < px.epos) {
      px.pos += 1;
      return true;
    }
    return false;
  };

  public static ParseFunc pAny() {
    return ANY;
  }

  public static ParseFunc pChar(String text) {
    final int text_length = text.length();
    return (px) -> {
      if (px.inputs.startsWith(text, px.pos)) {
        px.pos += text_length;
        return true;
      }
      return false;
    };
  }

  private static int find_codemax(String chars, String[] ranges) {
    int code = 0;
    for (int i = 0; i < chars.length(); i += 1) {
      code = Math.max(chars.charAt(i), code);
    }
    for (String range : ranges) {
      code = Math.max(range.charAt(0), code);
      code = Math.max(range.charAt(1), code);
    }
    return code;
  }

  private static void set_bitmap(byte[] bitmap, int c) {
    int n = (c / 8) | 0;
    int mask = 1 << ((c % 8) | 0);
    // console.log(n);
    // console.log(bitmap[n])
    bitmap[n] |= mask;
    // console.log(bitmap);
  }

  public static ParseFunc pRange(String chars, String[] ranges) {
    int codemax = find_codemax(chars, ranges) + 1;
    byte[] bitmap = new byte[((codemax / 8) | 0) + 1];
    bitmap[0] = 2;
    for (int i = 0; i < chars.length(); i += 1) {
      set_bitmap(bitmap, chars.charAt(i));
    }
    for (String range : ranges) {
      for (int c = range.charAt(0); c <= range.charAt(1); c += 1) {
        set_bitmap(bitmap, c);
      }
    }
    return (px) -> {
      if (px.pos < px.epos) {
        int c = px.inputs.charAt(px.pos);
        int n = (c / 8) | 0;
        int mask = 1 << ((c % 8) | 0);
        if (n < bitmap.length && (bitmap[n] & mask) == mask) {
          px.pos += 1;
          return true;
        }
      }
      return false;
    };
  }

  public static ParseFunc pMany(ParseFunc match) {
    return (px) -> {
      int pos = px.pos;
      ParTree ast = px.ast;
      while (match.apply(px) && px.pos > pos) {
        pos = px.pos;
        ast = px.ast;
      }
      px.head_pos = Math.max(px.pos, px.head_pos);
      px.pos = pos;
      px.ast = ast;
      return true;
    };
  }

  public static ParseFunc pMany1(ParseFunc match) {
    return (px) -> {
      if (match.apply(px)) {
        int pos = px.pos;
        ParTree ast = px.ast;
        while (match.apply(px) && px.pos > pos) {
          pos = px.pos;
          ast = px.ast;
        }
        px.head_pos = Math.max(px.pos, px.head_pos);
        px.pos = pos;
        px.ast = ast;
        return true;
      }
      return false;
    };
  }

  public static ParseFunc pAnd(ParseFunc match) {
    return (ParserContext px) -> {
      int pos = px.pos;
      if (match.apply(px)) {
        px.head_pos = Math.max(px.pos, px.head_pos);
        px.pos = pos;
        return true;
      }
      return false;
    };
  }

  public static ParseFunc pOr(ParseFunc match) {
    return (ParserContext px) -> {
      int pos = px.pos;
      ParTree ast = px.ast;
      if (match.apply(px)) {
        px.head_pos = Math.max(px.pos, px.head_pos);
        px.pos = pos;
        px.ast = ast;
        return false;
      }
      return true;
    };
  }

  public static ParseFunc pOption(ParseFunc match) {
    return (ParserContext px) -> {
      int pos = px.pos;
      ParTree ast = px.ast;
      if (!match.apply(px)) {
        px.head_pos = Math.max(px.pos, px.head_pos);
        px.pos = pos;
        px.ast = ast;
      }
      return true;
    };
  }

  public static ParseFunc pSeq(ParseFunc... matches) {
    return (ParserContext px) -> {
      for (ParseFunc match : matches) {
        if (!match.apply(px)) {
          return false;
        }
      }
      return true;
    };
  }

  public static ParseFunc pSeq2(ParseFunc match, ParseFunc match2) {
    return (ParserContext px) -> {
      return match.apply(px) && match2.apply(px);
    };
  }

  public static ParseFunc pSeq3(ParseFunc match, ParseFunc match2, ParseFunc match3) {
    return (ParserContext px) -> {
      return match.apply(px) && match2.apply(px) && match3.apply(px);
    };
  }

  public static ParseFunc pOre(ParseFunc... matches) {
    return (ParserContext px) -> {
      int pos = px.pos;
      ParTree ast = px.ast;
      for (ParseFunc match : matches) {
        if (match.apply(px)) {
          return true;
        }
        px.head_pos = Math.max(px.pos, px.head_pos);
        px.pos = pos;
        px.ast = ast;
      }
      return false;
    };
  }

  public static ParseFunc pOre2(ParseFunc match, ParseFunc match2) {
    return (ParserContext px) -> {
      int pos = px.pos;
      ParTree ast = px.ast;
      if (!match.apply(px)) {
        px.head_pos = Math.max(px.pos, px.head_pos);
        px.pos = pos;
        px.ast = ast;
        return match2.apply(px);
      }
      return true;
    };
  }

  public static ParseFunc pRef(HashMap<String, ParseFunc> peg, String name) {
    if (peg.containsKey(name)) {
      return peg.get(name);
    }
    return (ParserContext px) -> {
      return peg.get(name).apply(px);
    };
  }

  public static ParseFunc pNode(ParseFunc match, String tag, int shift) {
    return (ParserContext px) -> {
      int pos = px.pos;
      px.ast = null;
      if (match.apply(px)) {
        px.ast = new ParTree(null, tag, pos + shift, px.pos, px.ast);
        return true;
      }
      return false;
    };
  }

  public static ParseFunc pEdge(String edge, ParseFunc match) {
    return (ParserContext px) -> {
      ParTree ast = px.ast;
      if (match.apply(px)) {
        px.ast = new ParTree(ast, edge, -1, -1, px.ast);
        return true;
      }
      return false;
    };
  }

  public static ParseFunc pFold(String edge, ParseFunc match, String tag, int shift) {
    return (ParserContext px) -> {
      int pos = px.pos;
      px.ast = new ParTree(null, edge, -1, -1, px.ast);
      if (match.apply(px)) {
        px.ast = new ParTree(null, tag, pos + shift, px.pos, px.ast);
        return true;
      }
      return false;
    };
  }

  public static ParseFunc pAbs(String edge, ParseFunc match, String tag, int shift) {
    return (ParserContext px) -> {
      ParTree ast = px.ast;
      if (match.apply(px)) {
        px.ast = ast;
        return true;
      }
      return false;
    };
  }

  public static ParseFunc pSkipErr() {
    return (ParserContext px) -> {
      px.pos = Math.min(px.head_pos, px.epos);
      return true;
    };
  }

  public static ParseFunc pSymbol(int sid, ParseFunc match) {
    return (ParserContext px) -> {
      int pos = px.pos;
      if (match.apply(px)) {
        px.state = new State(sid, px.inputs.substring(pos, px.pos), px.state);
        return true;
      }
      return false;
    };
  }

  public static ParseFunc pExists(int sid, ParseFunc match) {
    return (ParserContext px) -> {
      return State.get(px.state, sid) != null;
    };
  }

  public static ParseFunc pMatch(int sid, ParseFunc match) {
    return (ParserContext px) -> {
      State state = State.get(px.state, sid);
      if (state != null) {
        if (px.inputs.startsWith(state.value, px.pos)) {
          px.pos += state.value.length();
          return true;
        }
      }
      return false;
    };
  }

  public static ParseFunc pScope(ParseFunc match) {
    return (ParserContext px) -> {
      State state = px.state;
      boolean res = match.apply(px);
      px.state = state;
      return res;
    };
  }

  static HashMap<String, ParseFunc> peg = null;

  static ParseFunc grammar(String start) {
    if (peg == null) {
      peg = new HashMap();
      // TPEG
    }
    return peg.get(start);
  }

  public static Function<String, ParTree> generate(String start) {
    ParseFunc match = grammar(start);
    if (match == null) {
      throw new RuntimeException("undefined " + start);
    }
    return (String inputs) -> {
      int pos = 0;
      ParserContext px = new ParserContext(null, inputs, 0, inputs.length());
      if (match.apply(px)) {
        if (px.ast == null) {
          px.ast = new ParTree(null, "", pos, px.pos, null);
        }
      } else {
        px.ast = new ParTree(null, "err", px.head_pos, px.head_pos + 1, null);
      }
      return px.ast;
    };
  }
}
